File: cfworker/src/middleware.ts
// Port of internal/server/middleware.go — the shared write-path behaviors,
// in the same order as the Go write() wrapper after its anonymous-claim
// address gate (applied by workspace-do.ts): per-user rate limit, then
// Idempotency-Key semantics (per principal, per endpoint — the endpoint is
// the exact route-pattern string — ≥24h retention; identical replay returns
// the original response, headers included; body mismatch is a 409;
// purge-on-write).

import type { ReqCtx } from "./context";
import type { Handler } from "./router";
import { capturedBody, problemResponse } from "./problems";
import { sha256Hex } from "./auth";
import { idemGet, idemPut } from "./store/idempotency";
import { userByTokenHash } from "./store/users";
import { runHandler, runHandlerSync } from "./handlers/helpers";
import type { RateLimiter } from "./ratelimit";

const RETENTION_MS = 24 * 60 * 60 * 1000;

export interface WritePath {
  limiter: RateLimiter;
  // One in-flight execution per (principal, endpoint, key): concurrent
  // retries of the same key serialize here, so the loser replays the
  // winner's response instead of double-executing. Between get and put
  // everything is synchronous SQL inside one transaction, which input gates
  // make non-interleavable anyway — the lock is kept as refactor insurance.
  withIdemLock<T>(key: string, f: () => Promise<T>): Promise<T>;
}

// principalFor identifies who a write is charged to: the bearer principal,
// or — for the unauthenticated claim endpoint — the username being claimed.
// Empty means unidentifiable (the handler's own auth will reject it). A
// deactivated user is deliberately unidentifiable too: their credential must
// not replay cached responses (which may hold issued secrets) — the handler's
// authenticate() rejects them with the spec'd 401 instead.
function principalFor(c: ReqCtx): string {
  if (c.tokenHash !== null) {
    const user = userByTokenHash(c.store, c.tokenHash);
    if (user === null || user.deactivated) return "";
    return user.username;
  }
  if (c.request.method === "POST" && c.url.pathname === "/v1/users") {
    try {
      const req = JSON.parse(c.bodyText) as { username?: unknown };
      if (req && typeof req.username === "string" && req.username !== "") {
        return "claim:" + req.username;
      }
    } catch {
      // unparseable body: unidentifiable
    }
  }
  return "";
}

export async function writeWrapped(w: WritePath, c: ReqCtx, endpoint: string, handler: Handler): Promise<Response> {
  const principal = principalFor(c);
  if (principal !== "") {
    const { ok, retryAfter } = w.limiter.allow(principal, Date.now());
    if (!ok) {
      return problemResponse(429, "rate-limited", "per-user write rate limit", {
        "Retry-After": String(retryAfter),
      });
    }
  }

  const key = c.request.headers.get("Idempotency-Key") ?? "";
  if (key === "" || principal === "") {
    return runHandler(handler, c);
  }
  if (key.length > 128) {
    return problemResponse(400, "validation", "Idempotency-Key over 128 characters");
  }

  const reqHash = await sha256Hex(c.bodyText);
  const lockKey = `${principal}\x00${endpoint}\x00${key}`;
  // Everything from the cache lookup through the handler's mutation to the
  // idempotency record is one transactionSync: the mutation and its
  // remembered result commit together, so a reset can never leave a
  // committed mutation that a retry would execute a second time. Write
  // handlers are synchronous by construction (all async work — body read,
  // token hashing/minting — happened in the DO's fetch), so no await can
  // escape the transaction.
  return w.withIdemLock(lockKey, async () =>
    c.store.tx(() => {
      const cutoff = Date.now() - RETENTION_MS;
      const rec = idemGet(c.store, principal, endpoint, key, cutoff);
      if (rec !== null) {
        if (rec.requestHash !== reqHash) {
          return problemResponse(
            409,
            "idempotency-key-conflict",
            "Idempotency-Key was already used with a different request body",
          );
        }
        return new Response(rec.body === "" ? null : rec.body, {
          status: rec.status,
          headers: rec.headers,
        });
      }

      const resp = runHandlerSync(handler, c);
      if (resp.status < 500) {
        const body = capturedBody(resp);
        if (body === null) {
          // Every handler builds responses via problems.ts, which captures
          // the body string; anything else is a programming error.
          throw new Error("write handler produced a response with no captured body");
        }
        idemPut(
          c.store,
          principal,
          endpoint,
          key,
          { requestHash: reqHash, status: resp.status, headers: [...resp.headers], body },
          Date.now(),
          cutoff,
        );
      }
      return resp;
    }),
  );
}

File: internal/server/middleware.go
package server

import (
	"bytes"
	"container/list"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"io"
	"math"
	"net/http"
	"strconv"
	"sync"
	"time"

	"github.com/dosu-ai/abbs/internal/store"
)

// limiter is an in-process token bucket keyed by the caller-selected identity
// (username for writes, observed address for anonymous reads and claims). No
// Redis until a second server node exists.
type limiter struct {
	mu         sync.Mutex
	burst      float64
	refill     float64 // tokens per second
	maxBuckets int
	buckets    map[string]*list.Element
	lru        *list.List
}

type bucket struct {
	key    string
	tokens float64
	last   time.Time
}

func newLimiter(burst int, refillPerSec float64, maxBuckets int) *limiter {
	return &limiter{
		burst: float64(burst), refill: refillPerSec, maxBuckets: maxBuckets,
		buckets: map[string]*list.Element{}, lru: list.New(),
	}
}

// allow consumes one token; when exhausted it reports the seconds to wait.
func (l *limiter) allow(user string, at time.Time) (ok bool, retryAfter int) {
	l.mu.Lock()
	defer l.mu.Unlock()
	elem, found := l.buckets[user]
	var b *bucket
	if !found {
		if l.lru.Len() >= l.maxBuckets {
			oldest := l.lru.Front()
			delete(l.buckets, oldest.Value.(*bucket).key)
			l.lru.Remove(oldest)
		}
		b = &bucket{key: user, tokens: l.burst, last: at}
		elem = l.lru.PushBack(b)
		l.buckets[user] = elem
	} else {
		b = elem.Value.(*bucket)
		l.lru.MoveToBack(elem)
	}
	b.tokens = min(l.burst, b.tokens+at.Sub(b.last).Seconds()*l.refill)
	b.last = at
	if b.tokens >= 1 {
		b.tokens--
		return true, 0
	}
	secs := int(math.Ceil((1 - b.tokens) / l.refill))
	if secs < 1 {
		secs = 1
	}
	return false, secs
}

// recorder captures a handler's response so it can be stored and replayed
// for idempotency-key retries.
type recorder struct {
	header http.Header
	status int
	body   bytes.Buffer
}

func newRecorder() *recorder { return &recorder{header: http.Header{}, status: http.StatusOK} }

func (r *recorder) Header() http.Header         { return r.header }
func (r *recorder) WriteHeader(status int)      { r.status = status }
func (r *recorder) Write(b []byte) (int, error) { return r.body.Write(b) }

func (r *recorder) flush(w http.ResponseWriter) {
	for k, vs := range r.header {
		for _, v := range vs {
			w.Header().Add(k, v)
		}
	}
	w.WriteHeader(r.status)
	w.Write(r.body.Bytes())
}

// write wraps a mutating handler with the write-path behaviors: the anonymous
// claim address limit, per-user rate limit, and Idempotency-Key semantics (per
// principal, per endpoint, ≥24h retention; identical replay returns the
// original response; body mismatch is a 409). endpoint is the route pattern —
// the spec's per-endpoint scope.
func (s *Server) write(endpoint string, handler http.HandlerFunc) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		// Read and restore the body: the principal for the claim endpoint,
		// the request hash, and the handler all need it.
		body, err := io.ReadAll(r.Body)
		if err != nil {
			writeProblem(w, http.StatusBadRequest, "validation", "cannot read request body")
			return
		}
		r.Body = io.NopCloser(bytes.NewReader(body))

		principal, activeBearer := s.principalFor(r, body)
		if endpoint == "POST /v1/users" && !activeBearer {
			if ok, retryAfter := s.claimLimiter.allow(s.anonymousClientKey(r), time.Now()); !ok {
				w.Header().Set("Retry-After", strconv.Itoa(retryAfter))
				writeProblem(w, http.StatusTooManyRequests, "rate-limited", "anonymous claim rate limit")
				return
			}
		}

		if principal != "" {
			if ok, retryAfter := s.limiter.allow(principal, time.Now()); !ok {
				w.Header().Set("Retry-After", strconv.Itoa(retryAfter))
				writeProblem(w, http.StatusTooManyRequests, "rate-limited", "per-user write rate limit")
				return
			}
		}

		key := r.Header.Get("Idempotency-Key")
		if key == "" || principal == "" {
			handler(w, r)
			return
		}
		if len(key) > 128 {
			writeProblem(w, http.StatusBadRequest, "validation", "Idempotency-Key over 128 characters")
			return
		}

		sum := sha256.Sum256(body)
		reqHash := hex.EncodeToString(sum[:])

		// One in-flight execution per key: concurrent retries of the same
		// key serialize here, so the loser replays the winner's response
		// instead of double-executing.
		lockKey := principal + "\x00" + endpoint + "\x00" + key
		mu, _ := s.idemLocks.LoadOrStore(lockKey, &sync.Mutex{})
		mu.(*sync.Mutex).Lock()
		defer mu.(*sync.Mutex).Unlock()

		cutoff := time.Now().Add(-24 * time.Hour)
		rec, err := s.store.IdemGet(principal, endpoint, key, cutoff)
		if err != nil {
			writeProblem(w, http.StatusInternalServerError, "internal", err.Error())
			return
		}
		if rec != nil {
			if rec.RequestHash != reqHash {
				writeProblem(w, http.StatusConflict, "idempotency-key-conflict",
					"Idempotency-Key was already used with a different request body")
				return
			}
			w.Header().Set("Content-Type", rec.ContentType)
			w.WriteHeader(rec.Status)
			w.Write(rec.Body)
			return
		}

		rr := newRecorder()
		handler(rr, r)
		rr.flush(w)
		if rr.status < 500 {
			_ = s.store.IdemPut(principal, endpoint, key, store.IdemRecord{
				RequestHash: reqHash,
				Status:      rr.status,
				ContentType: rr.header.Get("Content-Type"),
				Body:        rr.body.Bytes(),
			}, time.Now(), cutoff)
		}
	}
}

// principalFor identifies who a write is charged to and whether the request
// carries a bearer credential for an active user. The principal is the active
// bearer user, or — for the unauthenticated claim endpoint — the username
// being claimed. Empty means unidentifiable (the handler's own auth will
// reject it).
func (s *Server) principalFor(r *http.Request, body []byte) (principal string, activeBearer bool) {
	if token, ok := bearerToken(r); ok {
		if user, err := s.store.UserByTokenHash(hashToken(token)); err == nil && !user.Deactivated {
			return user.Username, true
		}
		return "", false
	}
	if r.Method == http.MethodPost && r.URL.Path == "/v1/users" {
		var req struct {
			Username string `json:"username"`
		}
		if json.Unmarshal(body, &req) == nil && req.Username != "" {
			return "claim:" + req.Username, false
		}
	}
	return "", false
}
